c_emulator/riscv_model_impl.h

```diff
@@ -0,0 +1,93 @@
+#pragma once
+
+#include <cstdint>
+#include <cstdio>
+#include <random>
+#include <vector>
+
+#include "sail.h"
+#include "sail_riscv_model.h"
+
+extern FILE *trace_log;
+extern int term_fd;
+void plat_term_write_impl(char c);
+
+// Model wrapped with an implementation of its platform callbacks.
+class ModelImpl : public model::Model {
```

Comment:
Might be worth marking this class as final so that the compiler can turn all the virtual function calls into a direct call. We can probably also change the way the sail emits C++ output, but using final should hopefully be equivalent when building with LTO.